client/components/popover/index.jsx

```diff
@@ -81,245 +86,176 @@ class Popover extends Component {
-			this.debug( 'no element to bind clickout ' );
-			return null;
-		}
-
-		if ( this._clickoutHandlerReference ) {
-			this.debug( 'clickout event already bound' );
-			return null;
-		}
-
-		this.debug( 'binding `clickout` event' );
-		this._clickoutHandlerReference = clickOutside( el, this.onClickout );
+	bindClickoutHandler() {
+		document.addEventListener( 'click', this.onClickout );
 	}
 
 	unbindClickoutHandler() {
-		if ( this._clickoutHandlerReference ) {
-			this.debug( 'unbinding `clickout` listener ...' );
-			this._clickoutHandlerReference();
-			this._clickoutHandlerReference = null;
-		}
+		document.removeEventListener( 'click', this.onClickout );
 	}
 
-	onClickout( event ) {
-		let shouldClose =
-			this.domContext && this.domContext.contains && ! this.domContext.contains( event.target );
+	onClickout = event => {
+		const popoverContext = this.popoverInnerNodeRef.current;
+		let shouldClose = popoverContext && ! popoverContext.contains( event.target );
+
+		if ( shouldClose && this.props.context ) {
+			const domContext = ReactDom.findDOMNode( this.props.context );
+			shouldClose = domContext && ! domContext.contains( event.target );
+		}
 
-		if ( this.props.ignoreContext && shouldClose ) {
+		if ( shouldClose && this.props.ignoreContext ) {
 			const ignoreContext = ReactDom.findDOMNode( this.props.ignoreContext );
-			shouldClose =
-				shouldClose &&
-				ignoreContext &&
-				ignoreContext.contains &&
-				! ignoreContext.contains( event.target );
+			shouldClose = ignoreContext && ! ignoreContext.contains( event.target );
 		}
 
 		if ( shouldClose ) {
 			this.close();
 		}
-	}
+	};
 
 	// --- window `scroll` and `resize` ---
-	bindDebouncedReposition() {
+	bindReposition() {
 		window.addEventListener( 'scroll', this.onWindowChange, true );
 		window.addEventListener( 'resize', this.onWindowChange, true );
 	}
 
-	unbindDebouncedReposition() {
-		if ( this.willReposition ) {
-			window.cancelAnimationFrame( this.willReposition );
-			this.willReposition = null;
-		}
-
+	unbindReposition() {
 		window.removeEventListener( 'scroll', this.onWindowChange, true );
 		window.removeEventListener( 'resize', this.onWindowChange, true );
-		this.debug( 'unbinding `debounce reposition` ...' );
-	}
-
-	onWindowChange() {
-		this.willReposition = window.requestAnimationFrame( this.setPosition );
-	}
-
-	focusPopover() {
-		if ( ! this.popoverNode ) {
-			return null;
-		}
-
-		this.debug( 'focusing the Popover' );
-		this.popoverNode.focus();
 	}
 
-	setDOMBehavior( domContainer ) {
-		if ( ! domContainer ) {
-			this.unbindClickoutHandler();
-			return null;
-		}
-
-		this.debug( 'setting DOM behavior' );
-
-		this.bindClickoutHandler( domContainer );
-
-		// store DOM element referencies
-		this.domContainer = domContainer;
-		this.popoverNode = document.getElementById( this.id );
-
-		// store context (target) reference into a property
-		this.domContext = ReactDom.findDOMNode( this.props.context );
+	onWindowChange = () => {
```

Comment:
`scroll` and `resize`, at least in modern browsers, fire at most once during an animation frame. So, `requestAnimationFrame` doesn't throttle anything. Are there browsers that fire more often than that? What debouncing delay would you recommend?

Reply:
Yes, you're right about modern browsers, they shouldn't emit events too often. To double-check, I tested Chrome, Safari, and Firefox, and none of them seem to emit the `resize` event more often than once per frame. Old browsers are generally the issue for these things.

However, we don't necessarily want to be emitting these events once per frame anyway. For resize, you're probably going to be blowing the frame budget anyway, as Calypso doesn't resize very smoothly. And JS handlers rarely go smoothly, too, at least unless they're passive. Chrome will probably activate an intervention and force the scroll handler to be passive, but as far as I know other browsers won't do the same, and it's not easy to feature-detect support for passive event listeners.

I tend to act defensively in generic, foundational components like this one, because you can never be sure what context they're going to be used in, how big of a sub-tree they're going to have, etc.

That said, I'll leave it to your judgment :) If you haven't noticed any runtime performance issues with the updated component then it's probably fine? Debouncing can be added in later if needed.

Reply:
When I try throttling or debouncing the scroll handler, it has some unpleasant visual effects: the popover repositioning is very visibly lagging behind the scroll:

![Screen Capture on 2019-11-21 at 18-38-02](https://user-images.githubusercontent.com/664258/69362368-ad439c00-0c8e-11ea-9cb6-1678e591434c.gif)

Therefore, I'd rather keep the synchronous callback, even though it can raise your CPU usage when scrolling. And work toward eliminating the repositioning handler completely in near future.

Reply:
Sounds good, thank you for looking into this! 👍